change(Actions): Major breaking rewrite of Actions

geolessel opened the pull request:
This completely changes the way the API works so we should really discuss. Here are some of the major points (I may have forgotten some as I started this rewrite a week or so ago):
- Split up `PcoApi.Actions` into submodules of `Get`, `Create`, `Self`, `New` (will also make `Update` when the time comes)
- Less magic in the modules for hitting the endpoints
- Split out a `list` function from `get`. `get` is now only responsible for getting singular records. `list` gets a list of records.
- When you define an endpoint module, you'll need to define just a handful of explicit functions:
  - `list\0` to define the listing URL
  - `list\1` to define the URL to use when passed a `Query`
  - `get\1` to define how to get a single record with an ID
  - `create\1` to define the URL used when creating a record
  - `self\1` to define the URL to use when a record needs to get itself but doesn't have a `self` link
  - `new\1` to define what kind of record this module is (ie `Person`, `Address`, etc)

Ignore all the endpoint modules expect `Person` in this PR if you can. I had to comment stuff out to get it to compile so I could test `Person`. Those tests pass, BTW.

What are your thoughts on this approach?

geolessel committed:
WIP

geolessel committed:
Pare down functions required in Person

jessejanderson commented on lib/pco_api/people/person.ex:
```diff
@@ -20,4 +19,15 @@ defmodule PcoApi.People.Person do
   linked_association :social_profiles
+
+  def list, do: get("people/v2/people")
+  def list(params) when is_list(params), do: get(params, "people/v2/people")
```
I know you do this in Actions already, but it seems like it might be nicer if `list/1` was just `def list, do: list([])` in this module to save you from needing the endpoint written twice here, makes it clear that they both end up at `get/2`. Not a big deal though.

geolessel replied:
I was actually thinking something similar as I was driving home from my daughter's ballet. That would make it one function defined per action. I like it.

jessejanderson commented on test/pco_api/people/person_test.exs:
```diff
@@ -10,51 +10,41 @@ defmodule PcoApi.People.PersonTest do
-    PcoApi.Query.where(first_name: "Geoffrey")
-    |> PcoApi.Query.where(last_name: "Lessel")
-    |> Person.get
+    assert [%PcoApi.Record{} | _rest] = Person.list
```
I think we should verify that `type` is `"Person"` in this test.

jessejanderson commented on lib/pco_api/people/person.ex:
```diff
@@ -20,4 +19,15 @@ defmodule PcoApi.People.Person do
+
+  def get(id) when is_integer(id), do: get("people/v2/people/#{id}")
+
+  def create(%PcoApi.Record{attributes: _, type: _} = record), do: create(record, "people/v2/people")
```
Feels kind of heavy to have to type so much to create a person.
`%PcoApi.Record{attributes: %{"first_name" => "geo", "last_name" => "lessel"}, type: "Person"} |> PcoApi.Person.create`

Can we just accept a map for creating a person providing the keys match the struct?
Then we could just do:
`PcoApi.Person.create(%{first_name: "geo", last_name: "lessel"})`

geolessel committed:
Bye-bye endpoint macro

geolessel commented:
@jessejanderson I went whole-hog and converted everything we've created up to this point to the new actions scheme. Not only that, but most things that have been completed for `.get` that have a corresponding create action, I've added that `.create` action with tests.

geolessel committed:
Merge branch 'master' into gl/actions-rewrite

geolessel committed:
Fix new tests

jessejanderson commented:
👍  for less macros.

Much love for `.list` over `.get`

credoci commented on lib/pco_api/actions/create.ex:
```diff
@@ -0,0 +1,33 @@
+        data |> to_record
+      end
+      defp do_create({:ok, %HTTPoison.Response{status_code: code, body: body}}) when (code in 200..299), do: body
+      defp do_create({:ok, %HTTPoison.Response{body: body}}), do: raise "PcoApi ok, but not ok: #{IO.inspect body}"
```
There should be no calls to IO.inspect/1.

credoci commented on lib/pco_api/actions/create.ex:
```diff
@@ -0,0 +1,33 @@
+      end
+      defp do_create({:ok, %HTTPoison.Response{status_code: code, body: body}}) when (code in 200..299), do: body
+      defp do_create({:ok, %HTTPoison.Response{body: body}}), do: raise "PcoApi ok, but not ok: #{IO.inspect body}"
+      defp do_create({:error, err}), do: raise "PcoApi error: #{IO.inspect err}"
```
There should be no calls to IO.inspect/1.

credoci commented on lib/pco_api/client.ex:
```diff
@@ -0,0 +1,34 @@
+      def request(:get, url, params) do
+        case get(url, [], params: params, hackney: [basic_auth: {PcoApi.key, PcoApi.secret}]) do
+          {:ok, %HTTPoison.Response{status_code: code, body: body}} when (code in 200..299) ->
+            cond do
```
Cond statements should contain at least two conditions besides `true`.

credoci commented on lib/pco_api/actions/new.ex:
```diff
@@ -0,0 +1,12 @@
+defmodule PcoApi.Actions.New do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/actions/self.ex:
```diff
@@ -0,0 +1,9 @@
+defmodule PcoApi.Actions.Self do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/actions/self.ex:
```diff
@@ -0,0 +1,9 @@
+    quote do
+      import PcoApi.Actions.Self
+
+      def self(%PcoApi.Record{links: %{"self" => self}}), do: get self
```
Parameter `self` has same name as a function in the same module.

credoci commented on test/pco_api/people/person_test.exs:
```diff
@@ -10,51 +10,41 @@ defmodule PcoApi.People.PersonTest do
-    person = Person.get(1)
-    assert %PcoApi.Record{id: "1"} = person
+    PcoApi.Query.where(first_name: "Geoffrey")
+    |> PcoApi.Query.where(last_name: "Lessel")
```
Pipe chain should start with a raw value.

credoci commented on lib/pco_api/actions/create.ex:
```diff
@@ -0,0 +1,33 @@
+      def post(url) when is_binary(url), do: create("", url)
+
+      def create(%PcoApi.Record{attributes: _, type: _} = record, url) when is_binary(url) do
+        # TODO: Error handling for when the record isn't created
```
Found a TODO tag in a comment: # TODO: Error handling for when the record isn't created

credoci commented on test/pco_api/people/email_test.exs:
```diff
@@ -0,0 +1,47 @@
+defmodule PcoApi.People.EmailTest do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/actions/get.ex:
```diff
@@ -0,0 +1,26 @@
+      def get(url, params) when is_binary(url) do
+        case get(url, [], params: params, hackney: [basic_auth: {PcoApi.key, PcoApi.secret}]) do
+          {:ok, %HTTPoison.Response{status_code: code, body: body}} when (code in 200..299) ->
+            cond do
```
Cond statements should contain at least two conditions besides `true`.

credoci commented on lib/pco_api/actions/get.ex:
```diff
@@ -0,0 +1,26 @@
+defmodule PcoApi.Actions.Get do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/actions/list.ex:
```diff
@@ -0,0 +1,10 @@
+defmodule PcoApi.Actions.List do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/actions.ex:
```diff
@@ -1,87 +1,27 @@
+      case unquote(opts) |> Keyword.has_key?(:only) do
+        true ->
+          unquote(opts) |> Keyword.fetch!(:only) |> Enum.each(fn(option) ->
+            case option do
```
Function body is nested too deep (max depth is 2, was 3).

credoci commented on lib/pco_api/actions/create.ex:
```diff
@@ -0,0 +1,33 @@
+defmodule PcoApi.Actions.Create do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/people/email.ex:
```diff
@@ -0,0 +1,24 @@
+defmodule PcoApi.People.Email do
```
Modules should have a @moduledoc tag.

credoci commented on test/pco_api/people/list_test.exs:
```diff
@@ -83,15 +83,26 @@ defmodule PcoApi.People.ListTest do
+      assert "POST" == conn.method
+      Plug.Conn.resp conn, 200, Fixture.dummy
+    end
+    record_with_links(bypass) |> List.run
```
Pipe chain should start with a raw value.

credoci commented on lib/pco_api/client.ex:
```diff
@@ -0,0 +1,34 @@
+defmodule PcoApi.Client do
```
Modules should have a @moduledoc tag.

credoci commented on lib/pco_api/record.ex:
```diff
@@ -5,4 +5,11 @@ defmodule PcoApi.Record do
   end
+
+  def to_record(results) when is_list(results) do
+    results |> Enum.map(&(&1 |> to_record))
```
Pipe chain should start with a raw value.
